### DISCUSSION TITLE
Dis-functioning Relationship  - or - how do you work with relationships?

### DISCUSSION THREAD

**Discussion by drorata** [fastapi/sqlmodel#612] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# models.py
from typing import List, Optional

from sqlmodel import Column, Field, Relationship, SQLModel, String

from foo_hardware import constants


class UserBase(SQLModel):
    username: str = Field(sa_column=Column("username", String, unique=True))
    password: str
    email: str = Field(sa_column=Column("email", String, unique=True))
    items: List["HardwareItem"] = Relationship(back_populates="owner")


class User(UserBase, table=True):
    id: int = Field(default=None, primary_key=True)


class UserCreate(UserBase):
    pass


class UserRead(UserBase):
    id: int


class HardwareItemBase(SQLModel):
    kind: constants.HARDWARE_KINDS = Field(
        default=...,
        title="Type of hardware",
        description=(
            "Possible kinds are:"
            f" {[(e.name, e.value) for e in constants.HARDWARE_KINDS]}"
        ),
    )
    owner_id: int = Field(foreign_key="user.id")
    owner: User = Relationship(back_populates="items")
    comment: Optional[str]


class HardwareItem(HardwareItemBase, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)


class HardwareItemCreate(HardwareItemBase):
    class Config:
        pass


class HardwareItemRead(HardwareItemBase):
    id: int


#########################
# backend.py
from typing import List

from fastapi import FastAPI, HTTPException, status
from sqlmodel import Session, SQLModel, create_engine, select

from foo_hardware import models

sqlite_file_name = "../data/foo_hardware.db"
sqlite_url = f"sqlite:///{sqlite_file_name}"

connect_args = {"check_same_thread": False}
engine = create_engine(sqlite_url, echo=True, connect_args=connect_args)


def create_db_and_tables():
    SQLModel.metadata.create_all(engine)


app = FastAPI()


@app.on_event("startup")
def on_startup():
    create_db_and_tables()


@app.post("/user/", response_model=models.UserRead)
def create_user(user: models.UserCreate):
    with Session(engine) as session:
        db_user = models.User.from_orm(user)
        session.add(db_user)
        session.commit()
        session.refresh(db_user)
        return db_user


@app.get("/user/", response_model=List[models.UserRead])
def read_users():
    with Session(engine) as session:
        users = session.exec(select(models.User)).all()
        return users


@app.post("/hardware/", response_model=models.HardwareItemRead)
def create_hardware_item(hardware_item: models.HardwareItemCreate):
    with Session(engine) as session:
        owner = session.exec(
            select(models.User).where(models.User.id == hardware_item.owner_id)
        ).fetchall()
        if len(owner) != 1:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND, detail="Owner not found"
            )
        owner = owner[0]
        db_hardware_item = models.HardwareItem.from_orm(hardware_item)
        # db_hardware_item.owner = owner
        session.exec
        session.add(db_hardware_item)
        session.commit()
        session.refresh(db_hardware_item)
        return db_hardware_item


@app.get("/hardware/", response_model=List[models.HardwareItemRead])
def read_hardware_items():
    with Session(engine) as session:
        hardware_items = session.exec(select(models.HardwareItem)).all()
        return hardware_items
```

### Description

The code above can also be found in [models.py](https://github.com/drorata/foo-hardware/blob/c8bbb1487d6c46657b1e977f8d6d9c59e5704361/foo_hardware/models.py) and [backend.py](https://github.com/drorata/foo-hardware/blob/c8bbb1487d6c46657b1e977f8d6d9c59e5704361/foo_hardware/backend.py).

It all works pretty much as expected, except [this](https://github.com/drorata/foo-hardware/blob/c8bbb1487d6c46657b1e977f8d6d9c59e5704361/foo_hardware/backend.py#L55-L57) part:

```python
        db_hardware_item = models.HardwareItem.from_orm(hardware_item)
        # db_hardware_item.owner = owner
        session.exec
```

As soon as I uncomment the middle line, I cannot add new hardware items. The error I get is something like:

```
ValueError: "HardwareItem" object has no field "owner"
```

which I find odd, because in [here](https://github.com/drorata/foo-hardware/blob/c8bbb1487d6c46657b1e977f8d6d9c59e5704361/foo_hardware/models.py#L37):

```python
owner: User = Relationship(back_populates="items")
```

it can be seen that `owner` is a member of `HardwareItemBase` which is the base class of `HardwareItem`.

I'm afraid I don't know how to solve this. I compared it a lot to the section [`relationship-attributes`](https://sqlmodel.tiangolo.com/tutorial/relationship-attributes/) in the docs, but the example there doesn't use the concept of `ItemBase`, `Item`, `ItemRead` and `ItemCreate` (which are discussed in [here](https://sqlmodel.tiangolo.com/tutorial/fastapi/multiple-models/?h=herocreate#the-herocreate-data-model) for example).

### Operating System

macOS

### SQLModel Version

0.0.8

### Python Version

3.11.2

**Comment by YuriiMotov** [comment]
I'll close this as it's old enough.
If you are still interested, please provide [MRE](https://stackoverflow.com/help/minimal-reproducible-example) and we will try to help
